### PR TITLE
Unit tests: reading binary UInt

### PR DESCRIPTION
**Pull request by zslayton**
*Issue #, if available:* #209 

*Description of changes:*

Adds `tests/IonParserBinaryRawTest.js`, which contains unit tests for:
* `ParserBinaryRaw::readUnsignedInt`
* `ParserBinaryRaw::readUnsignedLongInt`

This testing surfaced issue #210. UInts which require exactly 32 bits to represent currently cause read failures.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/IonParserBinaryRaw.ts [hunk 1]
```diff
@@ -155,13 +155,13 @@ export class ParserBinaryRaw {
         }
     }
 
-    private readVarUnsignedInt() : number {
+    static readVarUnsignedFrom(input: BinarySpan) : number {
```
Comment (zslayton) [inline src/IonParserBinaryRaw.ts]: The `readUnsignedInt` and `readVarUnsignedLongInt` methods both rely on internal state from the `ParserBinaryRaw` class. In order to make them testable, I've moved their core logic out into two new static methods: `readUnsignedIntFrom` and `readVarUnsignedLongIntFrom`. These methods allow you to bring your own state, making them easier to test.

File: src/IonParserBinaryRaw.ts [hunk 2]
```diff
@@ -213,28 +217,40 @@ export class ParserBinaryRaw {
...
-    private readUnsignedInt() : number {
-        var v = 0, b;
+    static readUnsignedIntFrom(input: BinarySpan, numberOfBytes: number) {
+        let value = 0, bytesRead = 0, byte;
+        if (numberOfBytes < 1)
+            return 0;
 
-        // they have to tell us the length
-        if (this._len < 1) return 0;
+        while (bytesRead < numberOfBytes) {
```
Comment (zslayton) [inline src/IonParserBinaryRaw.ts]: I expanded lots of single-character variable names into words for readability. The logic is the same, however.

File: src/IonParserBinaryRaw.ts [hunk 3]
```diff
@@ -213,28 +217,40 @@ export class ParserBinaryRaw {
...
-            v = v | (b & 0xff);
+        // We overflowed
+        if (value < 0) {
+            throw new Error("Attempted to read an unsigned int that was larger than 31 bits."
+                + " Use readUnsignedLongIntFrom instead. UInt size: " + numberOfBytes + ", value: " + value
+            );
         }
+        // Fewer bytes than the required `numberOfBytes` were available in the input
+        if (byte === EOF)
+            return undefined;
```
Comment (zslayton) [inline src/IonParserBinaryRaw.ts]: I believe that this should throw an `Exception` rather than returning `undefined`, but I'll leave that for another PR.

The original code accidentally omitted the `return` keyword, so it did nothing at all in this particular error case.

File: src/IonTests.ts [hunk 4]
```diff
@@ -20,4 +20,6 @@ export { Import } from './IonImport';
 export { LocalSymbolTable } from './IonLocalSymbolTable';
 export { LowLevelBinaryWriter } from './IonLowLevelBinaryWriter';
 export { NullNode } from './IonBinaryWriter';
-export { BinaryWriter } from "./IonBinaryWriter";
\ No newline at end of file
+export { BinaryWriter } from "./IonBinaryWriter";
+export { ParserBinaryRaw } from "./IonParserBinaryRaw";
```
Comment (zslayton) [inline src/IonTests.ts]: This makes the `ParserBinaryRaw` and `BinarySpan` classes accessible from within the unit test framework.

File: src/IonParserBinaryRaw.ts [hunk 5]
```diff
@@ -213,28 +217,40 @@ export class ParserBinaryRaw {
...
+        }
 
-        // shift in all but the last byte (we've already read the first)
-        while (this._len > 0) {
-            b = this._in.next();
-            this._len--;
-            v = v << 8;
-            v = v | (b & 0xff);
+        // We overflowed
+        if (value < 0) {
```
Comment (raganhan) [inline src/IonParserBinaryRaw.ts]: is there a chance to overflow twice so it's back to positive?

File: src/IonParserBinaryRaw.ts (hunk 3 above)
Comment (raganhan) [reply src/IonParserBinaryRaw.ts]: Agreed on the Exception. Is there an issue tracking this so it's not lost?

File: tests/unit/IonParserBinaryRawTest.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    };
+
+    // Registers the provided test with the test suite.
+    let registerTest = function(testName, test, throwsException) {
+      suite[testName] = throwsException
+          ? function() { assert.throws(test, Error); }
+          : test;
+    }
+
+    let readUnsignedIntTest = function(bytes, expected, throwsException) {
```
Comment (raganhan) [inline tests/unit/IonParserBinaryRawTest.js]: better to have two functions that call this one passing `true` or `false` for better test readability, e.g. `readUnsignedIntTest` and `readUnsignedIntTestReadMoreThanLength`

File: tests/unit/IonParserBinaryRawTest.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    readUnsignedIntTest([0x00], 0);
+    readUnsignedIntTest([0x01], 1);
+    readUnsignedIntTest([0x0F], 15);
+    readUnsignedIntTest([0x00, 0x00], 0);
+    readUnsignedIntTest([0xFF,], Math.pow(2, 8) - 1);
+    readUnsignedIntTest([0xFF, 0xFF], Math.pow(2, 16) - 1);
+    readUnsignedIntTest([0xFF, 0xFF, 0xFF], Math.pow(2, 24) - 1);
+    readUnsignedIntTest([0x7F, 0xFF, 0xFF, 0xFF], Math.pow(2, 31) - 1);
+
+    // Expected to fail
```
Comment (raganhan) [inline tests/unit/IonParserBinaryRawTest.js]: Add tests for overflow

File: src/IonParserBinaryRaw.ts (hunk 3 above)
Comment (zslayton) [reply src/IonParserBinaryRaw.ts]: I've gone ahead and turned it into an thrown `Error`.

File: src/IonParserBinaryRaw.ts (hunk 5 above)
Comment (zslayton) [reply src/IonParserBinaryRaw.ts]: In extreme cases, yes. I've modified the check to also fail if the number of bytes read was >4, accounting for all overflow cases beyond 32 bits.

File: tests/unit/IonParserBinaryRawTest.js (hunk 6 above)
Comment (zslayton) [reply tests/unit/IonParserBinaryRawTest.js]: I borrowed this from the other unit tests, but I like your approach much better. Done.

File: tests/unit/IonParserBinaryRawTest.js (hunk 7 above)
Comment (zslayton) [reply tests/unit/IonParserBinaryRawTest.js]: The tests immediately following check for overflow.
